keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,409 @@
+  not actually done so.
+
+2. That Kube-proxy provides a mechanism for load balancers to do connection
+  draining for terminating Nodes. This is to be done by Kube-proxy by inspecting
+  a field on the Node object which indicates that the "node is
+  terminating/deleting" and when seen, starts failing the HC. When discussing
+  this scenario, the primary case this applies to was: downscaling by the
+  cluster autoscaler (CA). Given that the CA taints the Node which is to
+  be downscaled and deleted, with the taint: `ToBeDeletedByClusterAutoscaler` it
+  would seem most appropriate to use that here. It is unfortunate to have this
+  taint spread around the code base, but for now: no better indicator has been
+  thought of.
+
+3. This KEP does not attempt to align cloud providers for health checking
+   `eTP:Cluster` services. It recognizes cloud providers have valid reasons for
+   doing this differently depending on their implementations. However, it does
+   want to recommend ways for cloud providers to do health checking in a better
+   way on Kubernetes clusters. The KEP hence proposes that a document be added
+   to https://kubernetes.io/docs/concepts/ which can act as a formal guide and
+   be utilized as knowledge sharing with cloud providers.
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Nodes used as intermediate nexthop for `eTP:Cluster` services would allow all
+  connections passing through the node while it is being terminated, to
+  gracefully shutdown. For `eTP:Local` workloads running on terminating nodes:
+  they get drained simultaneously with Kube-proxy today. This means that there's
+  no guarantee that connection draining will be enabled for them currently. That
+  might be the case if Kube-proxy is terminated before the endpoint, but there's
+  no guarantee of that. If Kube-proxy reacts to the
+  `ToBeDeletedByClusterAutoscaler` taint, then Kube-proxy can start prematurely
+  failing the HC and enable connection draining for these workloads.
+
+3. Cloud providers have very different ways of ascertaining if a load balancer
+   should target a specific node for `eTP:Cluster` services. We would like to
+   highlight the benefits of certain methods and pitfalls of others, in a formal
+   document, so that this is known. The hope is that this allows the information
+   to act as a source of knowledge for how to adapt their implementations to the
+   mechanics of a Kubernetes cluster.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for `eTP:Local` services.
```

Comment:
```suggestion
  for `eTP:Cluster` services.
```

is cluster , no?

Reply:
No, it's local. 

This happens today:
* for eTP:Cluster (at least on GCP): the LB HC targets Kube-proxy `healthz`
* for eTP:Local: the LB HC targets Kube-proxy on `healthCheckNodePort` and the answer doesn't reflect the `healthz` state. This is what this KEP proposes to change, that for eTP:Local we continue doing what we are doing but AND with `healthz`

Reply:
oh, ok , got it now, sorry I was confused because we were talking about the draining nodes problem all the time.

Regarding eTIP local and the `healthCheckNodePort`, after checking the code I think that eTP:local behavior is clearly a bug, if the healthzserver is never updated it keeps serving stale information.

 I suggest to fix the eTP problem directly and not mix it here, the KEP does a good work of explaining the draining problem with the Cluster behavior, IMHO we should not conflate both things cc: @thockin 

```
	if proxier.healthzServer != nil {
		proxier.healthzServer.Updated()
	}
	metrics.SyncProxyRulesLastTimestamp.SetToCurrentTime()

	// Update service healthchecks.  The endpoints list might include services that are
	// not "OnlyLocal", but the services list will not, and the serviceHealthServer
	// will just drop those endpoints.
	if err := proxier.serviceHealthServer.SyncServices(serviceUpdateResult.HCServiceNodePorts); err != nil {
		klog.ErrorS(err, "Error syncing healthcheck services")
	}
	if err := proxier.serviceHealthServer.SyncEndpoints(endpointUpdateResult.HCEndpointsLocalIPSize); err != nil {
		klog.ErrorS(err, "Error syncing healthcheck endpoints")
	}
```